### PR TITLE
Component System: Add FormGroup + ControlLabel components

### PR DESCRIPTION
**Pull request by ItsJonQ**
Screen Capture on 2021-01-28 at 16-14-19

This update adds a couple of low-level Form components for the [new Component System](https://github.com/WordPress/gutenberg/issues/27484), as part of https://github.com/WordPress/gutenberg/issues/28399

This 2 components are at the heart of creating form-based UI, such as those found with attribute controls / design tools:

Screen Shot 2021-01-28 at 4 34 20 PM

Compatible components (coming soon) would automatically be bound with labels (as well as `htmlFor` ids for a11y).
This is done through the internal `FormGroupContext`, resulting in a very minimal component API setup that looks like this:

```jsx
<FormGroup label="First name">
    <TextInput />
</FormGroup>
```

The example above would automatically render a `label` and  a unique `id`. That `id` would then be bound between the `label` (via `htmlFor`) and the `<TextInput />`.

In the [next Component System](https://github.com/ItsJonQ/g2/tree/main/packages/components/src), **all form based components** are pre-connected to this `FormGroupContext`, allowing them to work out of the box with zero setup.

## How has this been tested?

Locally in Storybook + Jest.
This component is already used in the next `FontSizePicker`, which has been tested in Gutenberg.

Local Storybook testing can found here:
http://localhost:50240/?path=/story/components-formgroup--default

After running `npm run storybook:dev`

## Types of changes

* Enhanced the `useInstanceId` hook to add a preferredId argument
* Moved over `ControlLabel` and `FormGroup` from [`@wp-g2/components`](https://github.com/ItsJonQ/g2/tree/main/packages/components/src)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by sarayourfriend — COMMENTED** [review]
LGTM! One last thing we need to do is add these components to the `packages/components/tsconfig.json` `include` rule.

**Comment by ItsJonQ** [comment]
@saramarcondes Haiii! I've added the files to `tsconfig.json` and did the best I could with types, but I'm getting destroyed, lol. 🙈

**Comment by saramarcondes** [comment]
@ItsJonQ Because `form-group` depends on `@wordpress/compose` which is untyped, we cannot type these modules. Anything that depends on `form-group` won't be able to be typed either. We should still add prop types to FormGroupContent/Help/Label however.

**Comment by ItsJonQ** [comment]
@saramarcondes Thanks for the heads up! The comment about `@wordpress/compose` makes sense to me.

Will do my best with the prop JSDoc type definitions

**Comment by ItsJonQ** [comment]
@saramarcondes Haii!! I've added types via JSDocs to the `FormGroup` components. I suspect I didn't do it correctly, haha.

I'm seeing this error:

> Syntax error in type: import('react').CSSProperties[ 'width' ]  jsdoc/valid-types

Any insight would be helpful! Thank you 🙏

**Comment by saramarcondes** [comment]
@ItsJonQ You did it just fine! It's just that the JSDoc parser that is used for our JSDoc linter doesn't understand complex types like function signatures and imports. If you check out other parts of the code based that are typed you'll notice that they're littered with `/* eslint-disable jsdoc/valid-types */`. This is unfortunate but I'm not sure what we can do about it other than disable the rule (but then we might start getting actually invalid types committed... who knows.

cc @sirreal @gziolo have y'all given any thought to disabling that rule? `jsdoc/no-undefined-types` is another that tends to give us trouble when typing JavaScript files.

**Comment by gziolo** [comment]
> @ItsJonQ You did it just fine! It's just that the JSDoc parser that is used for our JSDoc linter doesn't understand complex types like function signatures and imports. If you check out other parts of the code based that are typed you'll notice that they're littered with `/* eslint-disable jsdoc/valid-types */`. This is unfortunate but I'm not sure what we can do about it other than disable the rule (but then we might start getting actually invalid types committed... who knows.
> 
> cc @sirreal @gziolo have y'all given any thought to disabling that rule? `jsdoc/no-undefined-types` is another that tends to give us trouble when typing JavaScript files.

Yes, I noticed that for other G2 components as well. It looks like we are reaching the limit for those JSDoc rules. I would personally seek ways to disable them for those files that are typed with TypeScript. As far as I understand those two duplicate their work at the moment. We have `tsconfig.json` files in every package that has TS types defined, even when not fully, it could be a good compromise to use `glob` and disable those rules for packages that have at least one file typed. What do you think?

**Comment by saramarcondes** [comment]
@gziolo That sounds promising, however I'm not familiar with how to disable rules using file patterns in that way.

**Comment by sirreal** [comment]
When we're typechecking a package, the JSDoc type linter to be is totally irrelevant as far as I can tell. TypeScript tooling handles checking the types.

For packages that are typed or untyped, this is easy. The lint rule is on or off. The problem is that we have some packages now that are progressively being type, e.g. components. I don't think we want to disable the lint rule for these partially typed packages.

A good middle ground would be to disable it for the packages that are completely typed so they don't have to use so many eslint disables. Unfortunately, I think this would have to be a manual list. Doing anything more advanced like trying to parse the `tsconfig` files to see what's typechecked seems too complex for this problem.

**Comment by gziolo** [comment]
> A good middle ground would be to disable it for the packages that are completely typed so they don't have to use so many eslint disables. Unfortunately, I think this would have to be a manual list. Doing anything more advanced like trying to parse the `tsconfig` files to see what's typechecked seems too complex for this problem.

Can we use the following match as a way to disable those rules?
```json
"include": [ "src/**/*" ]
```

It would cover the case when some packages are partially typed. We would have to detect all such config files and use their parent folders to construct the rule for overrides. I can prototype something.

**Comment by sirreal** [comment]
> Can we use the following match as a way to disable those rules?
> 
> ```json
> "include": [ "src/**/*" ]
> ```

Looking at `include` in `tsconfig` is going to be complex, I believe. We might need to inspect `files`, `include`, and `exclude`. I may have a good idea for an alternative heuristic 🙂 

Our convention is to write types to the `dist-types` for publish. This requires a field to be added to the package.json when we're ready to publish the types, and we're only doing that when packages are fully typed. We may be able to check `package.json` for the presence of this, and disable the JSDoc type linter if it's present:

https://github.com/WordPress/gutenberg/blob/92baa9be10e45d91e5ce940600d3cb816630cac9/packages/a11y/package.json#L26

File: packages/components/src/form-group/test/index.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import ControlLabel from '../../control-label';
+import FormGroup, { useFormGroupContextId } from '../index';
+
+// @todo: Refactor this after adding next TextInput component.
+const TextInput = ( { id: idProp, ...props } ) => {
+	const id = useFormGroupContextId( idProp );
+	return <input type="text" id={ id } { ...props } />;
+};
+
+/* eslint-disable no-restricted-syntax */
```
Comment (gziolo) [inline packages/components/src/form-group/test/index.js:18]: What's the reason that this rule got disabled here?

**Comment by gziolo** [comment]
> Looking at `include` in `tsconfig` is going to be complex, I believe. We might need to inspect `files`, `include`, and `exclude`. I may have a good idea for an alternative heuristic 🙂
> 
> Our convention is to write types to the `dist-types` for publish. This requires a field to be added to the package.json when we're ready to publish the types, and we're only doing that when packages are fully typed. We may be able to check `package.json` for the presence of this, and disable the JSDoc type linter if it's present:
> 
> https://github.com/WordPress/gutenberg/blob/92baa9be10e45d91e5ce940600d3cb816630cac9/packages/a11y/package.json#L26

I opened https://github.com/WordPress/gutenberg/pull/28729 to address the issue with ESLint and TS types. It doesn't help with this PR just yet, but overall it should be a step in the right direction.

File: packages/components/src/form-group/test/index.js (hunk 1 above)
Comment (ItsJonQ) [reply packages/components/src/form-group/test/index.js:18]: It's mad at `id="fname"`

> Do not use string literals for IDs; use withInstanceId instead.

I can add a comment for that

**Comment by ItsJonQ** [comment]
With [VStack integrated](https://github.com/WordPress/gutenberg/pull/28798) (thanks @saramarcondes ), we can continue with this PR 🎉

**Comment by ItsJonQ** [comment]
I think tests are happy(ish) now! One of the E2E is still failing, but I don't think it's related to this update.

**Comment by ItsJonQ** [comment]
@gziolo / @sarayourfriend Haiii! Do you folks think we're 👍 on this one?

**Comment by ItsJonQ** [comment]
Local Storybook testing can found here:
http://localhost:50240/?path=/story/components-formgroup--default

After running `npm run storybook:dev`

**Comment by sarayourfriend** [comment]
Should we see if @diegohaz can do an a11y review for us? I think we've got our bases covered but just to make sure.

**Comment by ItsJonQ** [comment]
> ...@diegohaz can do an a11y review for us? I think we've got our bases covered but just to make sure.
@sarayourfriend Good idea ❤️

**Review by diegohaz — COMMENTED** [review]
This looks really good overall, but I noticed two issues with the help text:

1. The element should be associated with the input using `aria-describedby="help-text-id"` on the input.
2. The help text color doesn't meet the minimum contrast ratio required for WCAG 2.1 AA on a white background. You can check this with the Chrome Web Inspector:

    Web Inspector showing contrast ratio

    I guess `#767676` or darker should fix this.

**Comment by ItsJonQ** [comment]
@diegohaz Amazing! Thank you for checking 😍 

We have a design token for `muted` `<Text />` colors:
https://github.com/ItsJonQ/g2/blob/main/packages/styles/src/theme/config.js#L45

I can make the adjustment there, and it should ripple across the system

**Comment by ItsJonQ** [comment]
Created a PR here:
https://github.com/ItsJonQ/g2/pull/261

**Comment by sarayourfriend** [comment]
> I wonder if it's about time we bring in create-styles and styles into the Gutenberg repo @sarayourfriend 😉

I'm ready for this, but would want to check with @gziolo first to see if we're far enough along for it to be worth it. We have run into several times where we had to make updates and publish new versions of the style system though, so it'd be nice if those just lived in the `ui` folder.

**Comment by ItsJonQ** [comment]
> I'm ready for this, but would want to check with @gziolo first to see

That's fair :)

> so it'd be nice if those just lived in the ui folder.

Yessss 🙌

**Comment by ItsJonQ** [comment]
Screen Shot 2021-02-11 at 2 48 09 PM

Haii! With the recent push, the contrast for the `help` text in `FormGroup` should be fixed 👍

**Comment by ItsJonQ** [comment]
How are folks feeling with the latest updates :D

**Comment by gziolo** [comment]
> > I wonder if it's about time we bring in create-styles and styles into the Gutenberg repo @sarayourfriend 😉

> I'm ready for this, but would want to check with @gziolo first to see if we're far enough along for it to be worth it. We have run into several times where we had to make updates and publish new versions of the style system though, so it'd be nice if those just lived in the ui folder.

Do we have all components moved that `FontSizeControl` depends on? Let's make sure first that we don't reference `@wp-g2/components` in the `ui` folder anymore, and then we can immediately work on `create-styles` and `styles`. I agree that for start having those 2 packages as `ui/styles` and `ui/create-styles` subfolders would prevent some hassles with package and documentation wrangling 👍

**Comment by sarayourfriend** [comment]
@gziolo There are actually many left to go. All of the ones currently listed in #28399 are still required for the `FontSizeControl`.

**Comment by gziolo** [comment]
> There are actually many left to go. All of the ones currently listed in #28399 are still required for the `FontSizeControl`.

`Slider`, `Card` or `Menu` are necessary as well? Well, maybe, I assumed it's composed only of a few components 😄 You know better what to do, it doesn't matter that match what you will move first at this stage. I think it's easier to review components than the more advanced bits related to styles.

**Comment by sarayourfriend** [comment]
> Slider, Card or Menu are necessary as well

These are dependencies of other components that are used. Many components in the G2 system rely on context to render one way or another (`Divider` for example, depends on `Dropdown` for it's context, which in turn depends on [`Menu`](https://github.com/ItsJonQ/g2/blob/5feb4f2822efcd67afbf3f52849c78b8cbfd0f82/packages/components/src/Dropdown/DropdownMenu.js#L5)).

**Review by sarayourfriend — APPROVED** [review]
LGTM.

**Comment by gziolo** [comment]
> > Slider, Card or Menu are necessary as well
> 
> These are dependencies of other components that are used. Many components in the G2 system rely on context to render one way or another (`Divider` for example, depends on `Dropdown` for it's context, which in turn depends on [`Menu`](https://github.com/ItsJonQ/g2/blob/5feb4f2822efcd67afbf3f52849c78b8cbfd0f82/packages/components/src/Dropdown/DropdownMenu.js#L5)).

I find it a bit concerning. I hope that the `FontSizeControl` is an exception since it's very complex. Overall, I would assume that you should only need the code that is related to a given component so you could take advantage of tree-shaking when you use it with a 3rd party project that doesn't depend on WordPress. We should do an audit and verify whether the current dependencies in the code are optimal.

**Comment by sarayourfriend** [comment]
I think you're right that we're just hitting on the particularly complex parts of G2 with `FontSizeControl`. Tree shaking does work (https://github.com/ItsJonQ/g2/issues/37) but I guess there are some interconnected dependencies. They do not spider through the whole project though.

**Comment by diegohaz** [comment]
I'm not sure if it would work for G2, but one possible solution is to create more specific components when they depend on some other component's context. For example, `DropdownDivider` could depend on `Dropdown`, but not `Divider`.

**Pull request merged**